### DISCUSSION TITLE
Detecting board, but ignoring firmware version

### DISCUSSION THREAD

**Discussion by BeardOverflow** [BeardOverflow/msi-ec#26] [Ideas]
msi-ec is growing super fast (thanks a lot @teackot ) and a lot of new models are supported now.

Nowadays, to load a known configuration, we are probing the board identity and are linking it by ALLOWED_FW_* constants.

The board identity has the following pattern: xxxxxxxx.yyy

Some examples are:
- 14C1EMS1.012
- 14C1EMS1.101
- 14C1EMS1.102
- 17F2EMS1.103
- 17F2EMS1.104
- 17F2EMS1.106
- 17F2EMS1.107
- 1552EMS1.118
- 1592EMS1.111
- E1592IMS.10C
- 16V4EMS1.114

The xxxxxxxx pattern seems to be the real board name, while the yyy pattern is just the firmware version installed in the board. I think that it is safe to rely on the first pattern only. 

Pros:
- Easier to maintain by board name

Cons:
- A firmware update might change the embedded controller configuration, but it has never happened.

Finally, my question is: Has anyone found any incompatibility in this approach?

**Comment by teackot** [comment]
I was thinking about this too. Unfortunately, I found some incompatibilities when reading EC changelogs<sup>1</sup>

I found an incompatibility between <=17G1EMS1.105 and 17G1EMS1.107: in .107 a fan silence mode was added. That means it wasn't available before.

Also for 16W2EMS1:

> Note:
> If the old version is .3XX, choose this EC when updating.
> ...
> Note:
> If the old version is .1XX, choose this EC when updating.

.1yy and .3yy are uncompatible, but I don't know in what way as it isn't stated in the changelogs. Probably some hardware change that may or may not have affected EC addresses configuration.

What we can do for now is use version ranges instead of storing every single version as a string (yyy is a hexadecimal number). For example:

```
{
    board = "17G1EMS1",
    lower = 0,
    upper = 105,
},
{
    board = "17G1EMS1",
    lower = 107,
    upper = 107,
}
```

<sup>1</sup> you can get them from the official download page btw, [e.g.](https://www.msi.com/Laptop/GS75-Stealth-8SE/support#firmware)

**Comment by BeardOverflow** [comment]
Version ranges sounds better than current fixed-version approach. But, maybe it is not worth it because of are adding complexity to the code with minimal benefit :\

**Comment by teackot** [comment]
I agree. It isn't much faster and doesn't take too much memory (which is freed anyway because of __initdata (which should be replaced by __initconst btw))
